### PR TITLE
[feat] Provide cache hints via the data generation script

### PR DESCRIPTION
**Pull request by Gerson2102**
- [#141] issue #
- [X] follows contribution [guide](https://github.com/keep-starknet-strange/raito/blob/main/CONTRIBUTING.md)
- [ ] code change includes tests

Adding the cache hints for outpoint transactions. I did not test the code yet.

### CODE REVIEW

**Comment by Gerson2102** [comment]
I did not test the code yet, since I dont really know how to do it. Can someone help me with that a little bit?

**Comment by m-kus** [comment]
@maciejka adding linter would be definitely great!

@Gerson2102 all good, but you have to do another pass and update both outputs and outpoints in the `blocks` list.

**Comment by Gerson2102** [comment]
> @maciejka adding linter would be definitely great! I'm ok with the changes, we lacked consistency wrt quotes. But having this automated would help
> 
> @Gerson2102 all good, but you have to do another pass and update both outputs and outpoints in the `blocks` list.
> 
> How this will work on the Cairo side:
> 
> * When we handle transaction outputs and see cache hint, we add new item to the cache
> * When we handle transaction input and see cache hint in the outpoint, we try to pop item from the cache

So... I have to do some Cairo logic for this? Or just the python code? I dont get that part related to Cairo, or u just said it because I need to implement that on python?

**Comment by m-kus** [comment]
@Gerson2102 no need to implement Cairo logic here, it is just for the context so that you better understand what is it for. You just need to implement the Python part.

File: scripts/data/generate_data.py [hunk 1]
```diff
@@ -265,24 +265,49 @@ def generate_data(
             block = fetch_block_header(next_block_hash)
         elif mode == "full":
             block = fetch_block(next_block_hash)
+
             for tx in block["data"]:
                 for idx, output in enumerate(tx["outputs"]):
                     outpoint = (tx["txid"], idx)
                     utxo_set[outpoint] = output
 
+                    output["cached"] = False
```
Comment (m-kus) [inline scripts/data/generate_data.py:274]: This is already a default value

File: scripts/data/generate_data.py [hunk 2]
```diff
@@ -265,24 +265,49 @@ def generate_data(
...
+
+            for tx in block["data"]:
```
Comment (m-kus) [inline scripts/data/generate_data.py:276]: You can do in a single loop, but the order should be the following:
- First you iterate over all transaction inputs and set `cached=True` if the outpoint is in the utxo_set
- Then you iterate over all transaction outputs and add them to the utxo_set

File: scripts/data/generate_data.py [hunk 3]
```diff
@@ -265,24 +265,49 @@ def generate_data(
...
+    for block in blocks:
+        for tx in block["data"]:
+            for idx, output in enumerate(tx["outputs"]):
+                outpoint = (tx["txid"], idx)
+                if outpoint in utxo_set:
+                    tx["outputs"][idx]["cached"] = utxo_set[outpoint].get(
+                        "cached", False
+                    )
+
+            for tx_input in tx["inputs"]:
```
Comment (m-kus) [inline scripts/data/generate_data.py:301]: You already did that in L284

**Review by maciejka — APPROVED** [review]
LGTM, @m-kus

**Comment by maciejka** [comment]
@Gerson2102 fix conflicts, please.

**Comment by Gerson2102** [comment]
Conflicts fixed

File: scripts/data/generate_data.py (hunk 3 above)
Comment (m-kus) [reply scripts/data/generate_data.py]: @Gerson2102 but you still need to iterate over all outputs (after all the blocks are applied) to set `cache=True` if they have `utxo_set[outpoint]["cached"]`

To sum up, you need to do two passes:
1. Build utxo set, update inputs which consume cached outputs
2. Update outputs that are marked as "cached" in the utxo set

Comment (Gerson2102) [reply scripts/data/generate_data.py]: @m-kus check my last commit. This terms of outputs, outpoint, which I have to update is kinda confusing for me.

**Review by m-kus — APPROVED** [review]

**Pull request merged**